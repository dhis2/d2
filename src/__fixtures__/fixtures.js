export default (function x() {
    const fixtures = {};

    function getFixture(fixtureName) {
        if (fixtures && fixtures[fixtureName]) {
            // Make sure we return a new object
            return JSON.parse(JSON.stringify(fixtures[fixtureName]));
        }
        throw new Error(['Fixture', fixtureName, 'does not exist'].join(' '));
    }

    function addFixture(fixtureName, fixtureData) {
        if (fixtureName && fixtureData && typeof fixtureName === 'string') {
            fixtures[fixtureName] = fixtureData;
        }
    }

    addFixture('/api/attributes', require('./json/api/schemas/attributes.json'));
    addFixture('/api/dataElements', require('./json/api/dataElements.json'));
    addFixture('/api/organisationUnits', require('./json/api/organisationUnits.json'));
    addFixture('/api/schemas', require('./json/api/schemas.json'));
    addFixture('/api/schemas/dataElement', require('./json/api/schemas/dataElement.json'));
    addFixture('/api/schemas/dataSet', require('./json/api/schemas/dataSet'));
    addFixture('/api/schemas/indicatorGroup', require('./json/api/schemas/indicatorGroup.json'));
    addFixture('/api/schemas/organisationUnit', require('./json/api/schemas/organisationUnit.json'));
    addFixture('/api/schemas/organisationUnitGroupSet', require('./json/api/schemas/organisationUnitGroupSet'));
    addFixture('/api/schemas/validationRule', require('./json/api/schemas/validationRule'));
    addFixture('/api/schemas/program', require('./json/api/schemas/program'));
    addFixture('/api/schemas/programNotificationTemplate', require('./json/api/schemas/programNotificationTemplate'));
    addFixture('/api/schemas/user', require('./json/api/schemas/user.json'));
    addFixture('/api/schemas/legendSet', require('./json/api/schemas/legendSet.json'));
    addFixture('/api/schemas/legend', require('./json/api/schemas/legend.json'));
    addFixture('/api/userSettings', require('./json/api/userSettings'));
    addFixture('/appStore', require('./json/api/appStore.json'));
    addFixture('/dataElementAttributes', require('./json/dataElementAttributes.json'));
    addFixture('/me', require('./json/api/me.json'));
    addFixture('/me/authorities', require('./json/api/me/authorities.json'));
    addFixture('/modelDefinitions/dataElement', require('./json/modelDefinitions/dataElement.json'));
    addFixture('/singleUserAllFields', require('./json/singleUserAllFields.json'));
    addFixture('/singleUserOwnerFields', require('./json/singleUserOwnerFields.json'));
    addFixture('/api/legendSets/k1JHPfXsJND', require('./json/api/legendSets/k1JHPfXsJND.json'));
    addFixture('/api/analytics/dataValueSet', require('./json/api/analytics/dataValueSet.json'));
    addFixture('/api/analytics/rawData', require('./json/api/analytics/rawData.json'));
    addFixture('/api/analytics/aggregate', require('./json/api/analytics/aggregate.json'));
    addFixture('/api/analytics/count', require('./json/api/analytics/count.json'));
    addFixture('/api/analytics/query', require('./json/api/analytics/query.json'));
    addFixture('/api/analytics/cluster', require('./json/api/analytics/cluster.json'));

    return {
        get: getFixture,
        add: addFixture,
    };
}());
