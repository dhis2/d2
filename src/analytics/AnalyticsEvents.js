import AnalyticsBase from './AnalyticsBase';

/**
 * @extends module:analytics.AnalyticsBase
 *
 * @description
 * Analytics events class used to request analytics events data from Web API.
 *
 * @memberof module:analytics
 *
 * @see https://docs.dhis2.org/master/en/developer/html/webapi_event_analytics.html
 */
class AnalyticsEvents extends AnalyticsBase {
    /**
     * @param {!AnalyticsRequest} req Request object
     *
     * @returns {Promise} Promise that resolves with the analytics aggregate data from the api.
     *
     * @example
     * const req = new d2.analytics.request()
     *  .withProgram('eBAyeGv0exc')
     *  .addDataDimension(['Uvn6LCg7dVU','OdiHJayrsKo'])
     *  .addPeriodDimension('LAST_4_QUARTERS')
     *  .addOrgUnitDimension(['lc3eMKXaEfw','PMa2VCrupOd'])
     *  .addOrgUnitFilter('O6uvpzGd5pu')
     *  .withStartDate('2017-10-01')
     *  .withEndDate('2017-10-31');
     *
     *  d2.analytics.events.getAggregate(req)
     *  .then(console.log);
     */
    getAggregate(req) {
        return this.get(req.withPath('events/aggregate'));
    }

    /**
     * @param {!AnalyticsRequest} req Request object
     *
     * @returns {Promise} Promise that resolves with the analytics count data from the api.
     *
     * @example
     * const req = new d2.analytics.request()
     *  .withProgram('eBAyeGv0exc')
     *  .addDataDimension(['Uvn6LCg7dVU','OdiHJayrsKo'])
     *  .addPeriodDimension('LAST_4_QUARTERS')
     *  .addOrgUnitDimension(['lc3eMKXaEfw','PMa2VCrupOd'])
     *  .addOrgUnitFilter('O6uvpzGd5pu')
     *  .withStartDate('2017-10-01')
     *  .withEndDate('2017-10-31');
     *
     *  d2.analytics.events.getCount(req)
     *  .then(console.log);
     */
    getCount(req) {
        return this.get(req.withPath('events/count'));
    }

    /**
     * @param {!AnalyticsRequest} req Request object
     * Must contain clusterSize and bbox parameters.
     *
     * @returns {Promise} Promise that resolves with the analytics cluster data from the api.
     *
     * @example
     * const req = new d2.analytics.request()
     *  .withProgram('eBAyeGv0exc')
     *  .addDataDimension(['Uvn6LCg7dVU','OdiHJayrsKo'])
     *  .addPeriodDimension('LAST_4_QUARTERS')
     *  .addOrgUnitDimension(['lc3eMKXaEfw','PMa2VCrupOd'])
     *  .addOrgUnitFilter('O6uvpzGd5pu')
     *  .withStartDate('2017-10-01')
     *  .withEndDate('2017-10-31')
     *  .withClusterSize(100000)
     *  .withBbox('-13.2682125,7.3721619,-10.4261178,9.904012');
     *
     *  d2.analytics.events.getCluster(req)
     *  .then(console.log);
     */
    getCluster(req) {
        return this.get(req.withPath('events/cluster'));
    }

    /**
     * @param {!AnalyticsRequest} req Request object
     *
     * @returns {Promise} Promise that resolves with the analytics query data from the api.
     *
     * @example
     * const req = new d2.analytics.request()
     *  .withProgram('eBAyeGv0exc')
     *  .addDataDimension(['Uvn6LCg7dVU','OdiHJayrsKo'])
     *  .addPeriodDimension('LAST_4_QUARTERS')
     *  .addOrgUnitDimension(['lc3eMKXaEfw','PMa2VCrupOd'])
     *  .addOrgUnitFilter('O6uvpzGd5pu')
     *  .withStartDate('2017-10-01')
     *  .withEndDate('2017-10-31');
     *
     *  d2.analytics.events.getQuery(req)
     *  .then(console.log);
     */
    getQuery(req) {
        return this.get(req.withPath('events/query'));
    }
}

export default AnalyticsEvents;
