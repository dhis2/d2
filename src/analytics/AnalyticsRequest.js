import AnalyticsRequestDimensionsMixin from './AnalyticsRequestDimensionsMixin';
import AnalyticsRequestFiltersMixin from './AnalyticsRequestFiltersMixin';
import AnalyticsRequestPropertiesMixin from './AnalyticsRequestPropertiesMixin';
import AnalyticsRequestBase from './AnalyticsRequestBase';

/**
 * @description
 * Class for constructing a request object to use for communicating with the analytics API endpoint.
 *
 * @param {!Object} options Object with analytics request options
 *
 * @memberof module:analytics
 *
 * @extends module:analytics.AnalyticsRequestDimensionsMixin
 * @extends module:analytics.AnalyticsRequestFiltersMixin
 * @extends module:analytics.AnalyticsRequestPropertiesMixin
 * @extends module:analytics.AnalyticsRequestBase
 */
class AnalyticsRequest extends AnalyticsRequestDimensionsMixin(
    AnalyticsRequestFiltersMixin(AnalyticsRequestPropertiesMixin(AnalyticsRequestBase)),
) {
    /**
     * Extracts dimensions and filters from an analytic object model and add them to the request
     *
     * @param {Object} model The analytics object model from which extract the dimensions/filters
     *
     * @returns {AnalyticsRequest} A new instance of the class for chaining purposes
     *
     * @example
     * const req = new d2.analytics.request()
     *    .fromModel(model);
     *
     * // dimension=pe:LAST_12_MONTH&dimension=dx:fbfJHSPpUQD;cYeuwXTCPkU;Jtf34kNZhzP;hfdmMSPBgLG&filter=ou:ImspTQPwCqd
     */
    fromModel(model) {
        let request = this;

        // extract dimensions from model
        const columns = model.columns || [];
        const rows = model.rows || [];

        columns.concat(rows).forEach((d) => {
            let dimension = d.dimension;

            if (d.filter) {
                dimension += `:${d.filter}`;
            }

            request = request.addDimension(dimension, d.items.map(item => item.id));
        });

        // extract filters from model
        const filters = model.filters || [];

        filters.forEach(
            f => (request = request.addFilter(f.dimension, f.items.map(item => item.id))),
        );

        return request;
    }
}

export default AnalyticsRequest;
