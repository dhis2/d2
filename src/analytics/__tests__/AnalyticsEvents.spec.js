import fixtures from '../../__fixtures__/fixtures';
import MockApi from '../../api/Api';
import AnalyticsEvents from '../AnalyticsEvents';
import AnalyticsRequest from '../AnalyticsRequest';

jest.mock('../../api/Api'); // src/api/__mocks/Api.js

describe('analytics.events', () => {
    let events;
    let request;
    let mockApi;
    let fixture;

    beforeEach(() => {
        mockApi = MockApi.getApi();
        MockApi.mockClear();
        events = new AnalyticsEvents();
    });

    it('should not be allowed to be called without new', () => {
        expect(() => AnalyticsEvents()).toThrowError('Cannot call a class as a function'); // eslint-disable-line new-cap
    });

    it('should add the mockApi onto the Analyticsevents instance', () => {
        expect(events.api).toBe(mockApi);
    });

    it('should use the api object when it is passed', () => {
        const apiMockObject = {};

        events = new AnalyticsEvents(apiMockObject);

        expect(events.api).toBe(apiMockObject);
    });

    describe('.getAggregate()', () => {
        beforeEach(() => {
            events = new AnalyticsEvents(new MockApi());

            request = new AnalyticsRequest()
                .withLimit(10);

            fixture = fixtures.get('/api/analytics/aggregate');

            mockApi.get.mockReturnValue(Promise.resolve(fixture));
        });

        it('should be a function', () => {
            expect(events.getAggregate).toBeInstanceOf(Function);
        });

        it('should resolve a promise with data', () => events.getAggregate(request)
            .then((data) => {
                expect(data).toEqual(fixture);
            }));
    });

    describe('.getCount()', () => {
        beforeEach(() => {
            events = new AnalyticsEvents(new MockApi());

            request = new AnalyticsRequest()
                .withProgram('eBAyeGv0exc')
                .addPeriodDimension('LAST_YEAR')
                .addOrgUnitDimension('ImspTQPwCqd')
                .addDimension('qrur9Dvnyt5:LT:50');

            fixture = fixtures.get('/api/analytics/count');

            mockApi.get.mockReturnValue(Promise.resolve(fixture));
        });

        it('should be a function', () => {
            expect(events.getCount).toBeInstanceOf(Function);
        });

        it('should resolve a promise with data', () => events.getCount(request)
            .then((data) => {
                expect(data.count).toEqual(fixture.count);
                expect(data.extent).toEqual(fixture.extent);
            }));
    });

    describe('.getCluster()', () => {
        beforeEach(() => {
            events = new AnalyticsEvents(new MockApi());

            request = new AnalyticsRequest()
                .withProgram('VBqh0ynB2wv')
                .addOrgUnitDimension('ImspTQPwCqd')
                .withStage('pTo4uMt3xur')
                .withStartDate('2016-10-17')
                .withEndDate('2017-10-17')
                .withCoordinatesOnly(true)
                .withBbox('-14.062500000000002,5.61598581915534,-11.25,8.407168163601076')
                .withClusterSize(67265)
                .withIncludeClusterPoints(false);

            fixture = fixtures.get('/api/analytics/cluster');

            mockApi.get.mockReturnValue(Promise.resolve(fixture));
        });

        it('should be a function', () => {
            expect(events.getCluster).toBeInstanceOf(Function);
        });

        it('should resolve a promise with data', () => events.getCluster(request)
            .then((data) => {
                expect(data.width).toEqual(fixture.width);
                expect(data.height).toEqual(fixture.height);
            }));
    });

    describe('.getQuery()', () => {
        beforeEach(() => {
            events = new AnalyticsEvents(new MockApi());

            request = new AnalyticsRequest()
                .addOrgUnitDimension('ImspTQPwCqd')
                .addDimension('qrur9Dvnyt5:LT:50')
                .addPeriodFilter('LAST_MONTH')
                .withStage('Zj7UnCAulEk')
                .withPage(1)
                .withPageSize(5);

            fixture = fixtures.get('/api/analytics/query');

            mockApi.get.mockReturnValue(Promise.resolve(fixture));
        });

        it('should be a function', () => {
            expect(events.getQuery).toBeInstanceOf(Function);
        });

        it('should resolve a promise with data', () => events.getQuery(request)
            .then((data) => {
                expect(data.width).toEqual(fixture.width);
                expect(data.height).toEqual(fixture.height);
            }));
    });
});
