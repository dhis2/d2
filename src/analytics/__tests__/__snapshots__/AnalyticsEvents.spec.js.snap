// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analytics.events should not be allowed to be called without new 1`] = `"Class constructor AnalyticsEvents cannot be invoked without 'new'"`;
